### PR TITLE
Prototype improved a11y buffer view

### PR DESCRIPTION
**Pull request by Tyriar**
I got some feedback that the current navigation mode isn't very intuitive. This explores adding a different element, accessible via shift+tab (which will by default skip the terminal) that when focused fills a contenteditable element with the content of the buffer. This allows easy traversal using a screen reader and also selection.

image

An issue I'm currently hitting is when this element is created, NVDA can freeze for 15-60 seconds I'm guessing to update its model of the DOM. I tried caching the scrollback elements but that actually ends up worse when running `tree` as it doesn't do any debouncing, even if optimized it will probably add a noticeable delay between output and announcing. Capping the output to 100 is a workaround for this, it's not ideal though. A monaco instance may be a way to fix this problem, at least within VS Code.

I also thought we could leverage shell integration here in VS Code by exposing some delegate on the API that takes care of constructing the DOM elements, that way we could logically split up the buffer into command/output.

cc @meganrogge

### CODE REVIEW

File: src/browser/AccessibilityManager.ts [hunk 1]
```diff
@@ -83,7 +88,18 @@ export class AccessibilityManager extends Disposable {
     if (!this._terminal.element) {
       throw new Error('Cannot enable accessibility before Terminal.open');
     }
-    this._terminal.element.insertAdjacentElement('afterbegin', this._accessibilityTreeRoot);
+
+    const contentEditable = optionsService.options.accessibilityElementContentEditable;
```
Comment (meganrogge) [inline src/browser/AccessibilityManager.ts:92]: don't use content editable for now and remove this as an option

File: src/browser/AccessibilityManager.ts [hunk 2]
```diff
@@ -83,7 +88,18 @@ export class AccessibilityManager extends Disposable {
...
+    this._fullOutputElement = document.createElement(contentEditable ? 'section' : 'div');
+    this._fullOutputElement.contentEditable = contentEditable ? 'true' : 'false';
+
+    // TODO: Add to strings
```
Comment (meganrogge) [inline src/browser/AccessibilityManager.ts:96]: strings.ts localize

File: src/browser/AccessibilityManager.ts [hunk 3]
```diff
@@ -299,4 +320,41 @@ export class AccessibilityManager extends Disposable {
     this._liveRegion.textContent += this._charsToAnnounce;
     this._charsToAnnounce = '';
   }
+
+
+  private _refreshFullOutput(): void {
```
Comment (meganrogge) [inline src/browser/AccessibilityManager.ts:325]: change fullOutput to terminal accessibility mode? and css classes

File: src/browser/AccessibilityManager.ts [hunk 4]
```diff
@@ -26,6 +28,7 @@ export class AccessibilityManager extends Disposable {
   private _rowElements: HTMLElement[];
   private _liveRegion: HTMLElement;
   private _liveRegionLineCount: number = 0;
+  private _fullOutputElement: HTMLElement;
```
Comment (meganrogge) [inline src/browser/AccessibilityManager.ts:31]: accessible buffer

File: src/browser/AccessibilityManager.ts [hunk 5]
```diff
@@ -48,12 +51,14 @@ export class AccessibilityManager extends Disposable {
...
-    private readonly _renderService: IRenderService
+    @IOptionsService optionsService: IOptionsService,
+    @IRenderService private readonly _renderService: IRenderService,
+    @IThemeService themeService: IThemeService
   ) {
     super();
     this._accessibilityTreeRoot = document.createElement('div');
     this._accessibilityTreeRoot.classList.add('xterm-accessibility');
-    this._accessibilityTreeRoot.tabIndex = 0;
+    // this._accessibilityTreeRoot.tabIndex = 0;
```
Comment (meganrogge) [inline src/browser/AccessibilityManager.ts:61]: we need this

File: src/browser/AccessibilityManager.ts [hunk 6]
```diff
@@ -299,4 +320,41 @@ export class AccessibilityManager extends Disposable {
...
+    this._fullOutputElement.replaceChildren(...bufferElements);
+    const s = document.getSelection();
+    if (s && cursorElement) {
+      s.removeAllRanges();
+      const r = document.createRange();
+      r.selectNode(bufferElements[bufferElements.length - 1]);
+      r.setStart(cursorElement, 0);
+      r.setEnd(cursorElement, 0);
+      s.addRange(r);
+    }
```
Comment (meganrogge) [inline src/browser/AccessibilityManager.ts:347]: can rm

**Review by Tyriar — COMMENTED** [review]
Some lint issues:

```
/home/vsts/work/1/s/src/browser/AccessibilityManager.ts
  333:50  warning  Strings must use singlequote  @typescript-eslint/quotes

/home/vsts/work/1/s/src/browser/TestUtils.test.ts
  143:110  warning  Expected a comma          @typescript-eslint/member-delimiter-style
  143:151  warning  Unexpected separator (;)  @typescript-eslint/member-delimiter-style
  316:110  warning  Expected a comma          @typescript-eslint/member-delimiter-style
  316:151  warning  Unexpected separator (;)  @typescript-eslint/member-delimiter-style

/home/vsts/work/1/s/src/browser/Types.d.ts
  147:91  warning  Expected a comma  @typescript-eslint/member-delimiter-style

/home/vsts/work/1/s/src/browser/Viewport.ts
  283:98  warning  Expected a comma  @typescript-eslint/member-delimiter-style

✖ 7 problems (0 errors, 7 warnings)
  0 errors and 7 warnings potentially fixable with the `--fix` option.
```

File: src/browser/AccessibilityManager.ts [hunk 7]
```diff
@@ -83,7 +88,14 @@ export class AccessibilityManager extends Disposable {
     if (!this._terminal.element) {
       throw new Error('Cannot enable accessibility before Terminal.open');
     }
-    this._terminal.element.insertAdjacentElement('afterbegin', this._accessibilityTreeRoot);
+
+    this._accessibleBuffer = document.createElement('div');
+    this._accessibleBuffer.ariaLabel = Strings.accessibleBuffer;
+    this._accessibleBuffer.classList.add('xterm-accessibility-buffer');
+    this._refreshAccessibilityBuffer();
```
Comment (Tyriar) [inline src/browser/AccessibilityManager.ts:95]: A `// HACK:` comment saying this was needed to get it to work with a screen reader would be good for the future

File: src/browser/AccessibilityManager.ts [hunk 8]
```diff
@@ -299,4 +316,33 @@ export class AccessibilityManager extends Disposable {
...
+      return;
+    }
+
+    const { bufferElements, cursorElement } = this._terminal.viewport.getBufferElements(0);
+    for (const element of bufferElements) {
+      if (element.textContent) {
+        element.textContent = element.textContent.replace(new RegExp(' ', 'g'), '\xA0');
+      }
+    }
+    this._accessibleBuffer.tabIndex = 0;
```
[the remark this answers is not in the document]
Comment (meganrogge) [reply src/browser/AccessibilityManager.ts]: created #4373 and will seek feedback when this is merged and users can test

File: src/browser/AccessibilityManager.ts [hunk 9]
```diff
@@ -299,4 +316,33 @@ export class AccessibilityManager extends Disposable {
...
+  }
+
+  private _handleColorChange(colorSet: ReadonlyColorSet): void {
+    this._accessibleBuffer.style.backgroundColor = colorSet.background.css;
+    this._accessibleBuffer.style.color = colorSet.foreground.css;
+  }
+
+  private _handleFontOptionChange(options: Required<ITerminalOptions>): void {
+    this._accessibleBuffer.style.fontFamily = options.fontFamily;
+    this._accessibleBuffer.style.fontSize = `${options.fontSize}px`;
```
[the remark this answers is not in the document]
Comment (meganrogge) [reply src/browser/AccessibilityManager.ts]: created #4374

File: src/browser/LocalizableStrings.ts [hunk 10]
```diff
@@ -10,3 +10,5 @@ export let promptLabel = 'Terminal input';
 
 // eslint-disable-next-line prefer-const
 export let tooMuchOutput = 'Too much output to announce, navigate to rows manually to read';
+
+export const accessibilityBuffer = 'Accessibility buffer';
```
Comment (Tyriar) [inline src/browser/LocalizableStrings.ts:14]: This needs to be added to xterm.d.ts too https://github.com/Tyriar/xterm.js/blob/9ea1cd34d2d15cb5f7e1e4b34512fa52027c5c11/typings/xterm.d.ts#L490

**Review by Tyriar — COMMENTED** [review]
One more lint issue

**Comment by meganrogge** [comment]
fixed the lint issue, will add that

**Comment by JasonXJ** [comment]
Hi, I wonder what is the current status of this? It seems that on ToT, the old a11y tree based on list elements is gone, but the new thing (based on contexteditable element?) is also not ready? I am not able to use the screen reader on the terminal any more.

**Comment by Tyriar** [comment]
@JasonXJ the situation is vscode should be much better now, for other xterm.js embedders we also brought back the old a11y tree.

**Pull request merged**